### PR TITLE
Win console

### PR DESCRIPTION
**Pull request by daschuer**
This Pull Request hopefully allows to build mixxx and mixxx-test with subsystem:Windows. 
The console is created via custom code after main. 

The build flag DEBUGCONSOLE was also removed since Mixxx for windows should now work almost similar as Linux.

Related Bug:
https://bugs.launchpad.net/mixxx/+bug/1453782

Since I cannot build on windows this PR is untested.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -268,23 +269,23 @@ int main(int argc, char * argv[])
...
-    if (GetConsoleWindow() != NULL) {
-        // This should create the window for the hidden console
-        // initalized by Qt
-        SetConsoleTitleA("Mixxx Debug Messages");
-    }
-#endif
     UINT oldCodePage;
     bool shouldResetCodePage = false;
-    if (GetConsoleWindow() != NULL) {
+    if(AttachConsole(ATTACH_PARENT_PROCESS)) {
```
Comment (rryan) [inline src/main.cpp]: please pull all of this out into a helper function in util/console.h or similar to prevent duplication

File: src/SConscript [hunk 2]
```diff
@@ -94,19 +94,12 @@ def build_tests():
...
                 # both environments which makes SCons unhappy.
-                # # We want a terminal for tests.
-                # if build.machine_is_64bit:
-                #     test_env['LINKFLAGS'].remove('/subsystem:windows,5.02')
-                #     test_env['LINKFLAGS'].append('/subsystem:console,5.02')
-                # else:
-                #     test_env['LINKFLAGS'].remove('/subsystem:windows,5.01')
-                #     test_env['LINKFLAGS'].append('/subsystem:console,5.01')
-
+                # Currently booth Exes ar build with /subsystem:windows,5.02
```
Comment (rryan) [inline src/SConscript]: typos: "Currently both executables are built with /subsystem:windows and the console is attached manually"

**Comment by daschuer** [comment]
Done! Looking forward to see if it builds.

File: src/util/console.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    // In case of a German Windows XP to 10 console encoding is cp850
+    // where files encoding is cp1252
+    // Qt has no solution for it https://bugreports.qt.io/browse/QTBUG-13303
+    // http://stackoverflow.com/questions/1259084/what-encoding-code-page-is-cmd-exe-using
+    // We try to change the console encoding to file encoding
+    // For a German windows we expect
+    // LOCALE_IDEFAULTANSICODEPAGE "1252" // ANSI Codepage used by Qt toLocal8Bit
+    // LOCALE_IDEFAULTCODEPAGE "850" // OEM Codepage Console
+
+    UINT m_oldCodePage;
```
Comment (rryan) [inline src/util/console.cpp]: local shadows member variable

File: src/util/console.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
+
+#include "console.h"
```
Comment (rryan) [inline src/util/console.cpp]: Include QtDebug for qWarning

File: src/util/console.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    UINT m_oldCodePage;
+    bool m_shouldResetCodePage = false;
+    if(AttachConsole(ATTACH_PARENT_PROCESS)) {
+        // we are started from a console porcess
+        int fd;
+        FILE *fp;
+
+        fd = _open_osfhandle((long) GetStdHandle(STD_OUTPUT_HANDLE), 0);
+        fp = _fdopen(fd, "w");
+        *stdout = *fp;
```
Comment (rryan) [inline src/util/console.cpp]: I believe you need stdio.h for stdin/stderr. It may be included from other headers though.

Comment (rryan) [inline src/util/console.cpp]: https://msdn.microsoft.com/en-us/library/3x292kth.aspx

**Comment by daschuer** [comment]
fixed

**Comment by default-kramer** [comment]
Hey I'm back, sorry for the delay. These changes look good, but for some reason I don't have console output anymore on Windows 7. The only way I can see stdout and stderr is to pipe to a file. All the Qt logging is visible in stderr, so I personally don't think this is a problem.

I don't see how this PR could have caused this change in behavior, but I don't know what else it could be. If I type "dist32\mixxx.exe" on the command prompt it behaves as if I had typed "start dist32\mixxx.exe". In other words, Mixxx seems disconnected from the console immediately, before the code even has a chance to run. Did something change with the build process or Qt configuration maybe?

**Comment by daschuer** [comment]
> These changes look good, but for some reason I don't have console output anymore on Windows 7

So It seams I have failed :-( I can confirm this on WinXp: No console when starting in cmd.exe. 
Can you take care of it? According to what I have read it should work. So Maybe there is just a bug in the code? 
Thank you.

**Comment by default-kramer** [comment]
Ah, I see it now. This https://github.com/mixxxdj/mixxx/commit/304e5f6eceb60c8dbc80999f52f03c9876d6b577 change is most likely responsible. I will uncomment and try it.

**Comment by rryan** [comment]
@default-kramer -- indeed --  that was part of the plan. Remove subsystem:console from mixxx-test.exe and see if the console changes made recently cause the console to work.

**Comment by rryan** [comment]
The way the build is currently structured makes it hard to build mixxx.exe and mixxx-test.exe with different subsystem arguments. It's probably do-able though.

**Comment by rryan** [comment]
Also, mixxx-test output works on the build server despite subsystem:windows: http://builds.mixxx.org/jenkins/job/1.12-release/architecture=i386,platform=windows/34/console

**Comment by default-kramer** [comment]
When I switch back to subsystem:console, the console reappears as expected. But it always appears, even when Mixxx is not launched from a console. I think there is a way to conditionally hide it, although it may still appear in the taskbar which would probably not be acceptable.

I don't think it is possible to use subsystem:windows and attach to the console that launched Mixxx. The console is not blocked, so I could be typing in another command. I don't think that Mixxx would be allowed to hijack a console that belongs to the user and start spewing output while the user is trying to enter another command.

To summarize, I think the choice is between:

1. subsystem:console with conditionally hiding the console
2. subsystem:windows with conditionally creating a new console

Would either or both of these be acceptable?

**Comment by daschuer** [comment]
1. Has the disadvantage that the console will be shown at least for a spit of a second. So 2.) should be our favorite choice. 

@rryan: What if you start mixxx-test.exe inside cmd.exe? Do you see the output? 

@default-kramer Dos this example show a console output fromcmd.exe when compiled as subsystem windows? 
http://www.tillett.info/2013/05/13/how-to-create-a-windows-program-that-works-as-both-as-a-gui-and-console-application/

There is a problem anyway: we need to put the Console() call above the command line explanations to have them on the console. Maybe it fixes the problem at all?

**Comment by default-kramer** [comment]
@daschuer I had the same thought and tried moving Console() to the very beginning of main() but it didn't work. However, the example you linked to does work! So I should be able to get it working now.

**Comment by daschuer** [comment]
Cool! Thank you for taking care.
